migratekvdb/migration.go

```diff
@@ -0,0 +1,1296 @@
+	if m.migration.persistedState.ChunkSize != m.cfg.ChunkSize {
+		m.cfg.Logger.Error("Chunk size mismatch, previous migration "+
+			"chunk size was %d, but current chunk size is %d",
+			m.migration.persistedState.ChunkSize, m.cfg.ChunkSize)
+
+		return errChunkSizeMismatch
+	}
+
+	// Otherwise we are ready to resume the migration.
+	m.migration.resuming = true
+	m.cfg.Logger.InfoS(ctx, "Resuming migration",
+		"bucket", m.migration.persistedState.CurrentBucketPath,
+		"processed_keys", m.migration.persistedState.ProcessedKeys,
+	)
+
+	return nil
+}
+
+// Migrate performs the migration of the source database to the target database.
+func (m *Migrator) Migrate(ctx context.Context, sourceDB,
+	targetDB walletdb.DB) error {
+
+	m.cfg.Logger.Infof("Migrating database ...")
+
+	// Initialize or resume migration state.
+	if err := m.initMigrationState(ctx, targetDB); err != nil {
+		if err == errMigrationComplete {
+			return nil
+		}
+		return err
+	}
+
+	// Process chunks until complete.
+	for {
+		// Each chunk is processed in its own transaction pair.
+		needsMoreChunks, err := m.processNextChunk(
+			ctx, sourceDB, targetDB,
+		)
+		if err != nil {
+			return err
+		}
+
+		// Migration is complete.
+		if !needsMoreChunks {
+			return nil
+		}
+
+		// We start a new chunk.
+		m.migration.newChunk()
+	}
+}
+
+// processNextChunk processes the next chunk of the migration.
+func (m *Migrator) processNextChunk(ctx context.Context, sourceDB,
+	targetDB walletdb.DB) (bool, error) {
+
+	var needsMoreChunks bool
+
+	err := walletdb.View(sourceDB, func(sourceTx walletdb.ReadTx) error {
+		return walletdb.Update(targetDB, func(targetTx walletdb.ReadWriteTx) error {
+			// Process the chunk.
+			err := m.processChunk(ctx, sourceTx, targetTx)
+			if err != nil {
+				if err != errChunkSizeExceeded {
+					return err
+				}
+			}
+
+			if err == errChunkSizeExceeded {
+				// Save state and hash for the chunk.
+				err = m.migration.write(targetTx)
+				if err != nil {
+					return fmt.Errorf("failed to save "+
+						"migration state: %w", err)
+				}
+
+				err = m.recordChunkHash(ctx, targetTx)
+				if err != nil {
+					return fmt.Errorf("failed to record "+
+						"chunk hash: %w", err)
+				}
+
+				m.cfg.Logger.InfoS(
+					ctx, "Committed chunk successfully:",
+					"bucket", m.migration.persistedState.CurrentBucketPath,
+					"processed_keys", m.migration.persistedState.ProcessedKeys,
+					"current_chunk_size(B)", m.migration.currentChunkBytes,
+					"source_hash", m.migration.hash.Sum64(),
+				)
+
+				m.cfg.Logger.Infof("Migration progress saved, " +
+					"continuing with next chunk")
+
+				needsMoreChunks = true
+
+				return nil
+			}
+
+			// Migration is complete, finalize everything in this
+			// transaction.
+			m.cfg.Logger.Infof("Migration complete, processed in "+
+				"total %d keys",
+				m.migration.persistedState.ProcessedKeys)
+
+			// Mark migration as complete.
+			err = m.markMigrationComplete(targetTx)
```

Comment:
Shouldn't this only be done once the we have no more chunks to migrate?

Reply:
yes changed now.